File: coral-core/src/main/java/com/bhinneka/coral/core/EventStore.java
package com.bhinneka.coral.core;

import com.bhinneka.coral.core.handlers.AsyncHandler;

import java.util.List;

/**
 * Implement this store to persist and load {@link EventInfo EventInfo}.
 *
 * @param <I> entity id.
 * @param <S> the state of the entity.
 */
public interface EventStore<I, S> {

  /**
   * Loads all events for and entity of specific id.
   *
   * @param id the entity id to query
   * @param handler the handler of the load operation
   */
  void load(I id, AsyncHandler<List<EventInfo<I, S>>> handler);

  /**
   * Loads all events of an entity to specific version.
   *
   * @param id the entity id.
   * @param version the last version queried.
   * @param handler the handler of load operation.
   */
  void load(I id, int version, AsyncHandler<List<EventInfo<I, S>>> handler);

  /**
   * Commits an event to the event store.
   *  @param eventInfo event information to be persisted.
   * @param handler the handler of the commit operation
   */
  void commit(EventInfo<I, S> eventInfo, AsyncHandler<I> handler);
}

File: coral-kotlin/src/main/kotlin/com/ykode/coral/core/EventStore.kt
package com.ykode.coral.core

import com.ykode.coral.core.handlers.AsyncHandler;

/**
 * Event store load extension
 *
 * @param I the Entity ID type
 * @param S the state type
 * @param id the entity id to be loaded
 * @param success the success callback
 * @prama error the error callback
 */
fun <I, S> EventStore<I, S>.load(id: I, success: (List<EventInfo<I, S>>) -> Unit, error: (Throwable) -> Unit) =
  this.load(id, object : AsyncHandler<List<EventInfo<I, S>>> {
    override fun onError(exception: Exception) = error(exception)
    override fun onSuccess(result: List<EventInfo<I, S>>) = success(result)
  })

fun <I, S> EventStore<I, S>.load(id: I, version: Int, success: (List<EventInfo<I, S>>) -> Unit, error: (Throwable) -> Unit) =
    this.load(id, version, object : AsyncHandler<List<EventInfo<I, S>>> {
      override fun onError(exception: Exception) = error(exception)
      override fun onSuccess(result: List<EventInfo<I, S>>) = success(result)
    })

fun <I, S> EventStore<I, S>.commit(info: EventInfo<I, S>, success: (I) -> Unit, error: (Throwable) -> Unit) =
    this.commit(info, object : AsyncHandler<I> {
      override fun onError(exception: Exception) = error(exception)
      override fun onSuccess(result: I) = success(result)
    })